cassdegrees/static/js/rules.js

```diff
@@ -0,0 +1,226 @@
+//! Vue.js based means of adding/removing rules. Excludes serialization (see programmanagement.js)
+
+// Translation table between internal names for components and human readable ones.
+const COMPONENT_NAMES = {
+    'subplan': "Subplan",
+    'custom_text': "Custom (Text)"
+};
+
+Vue.component('rule_subplan', {
+    props: {
+        "details": {
+            type: Object,
+
+            validator: function (value) {
+                // Ensure that the object has all the attributes we need
+                if (!value.hasOwnProperty("ids")) {
+                    value.ids = [-1];
+                }
+
+                return true;
+            }
+        }
+    },
+    data: function() {
+        return {
+            "subplans": [],
+
+            // Display related warnings if true
+            "non_unique_options": false,
+            "inconsistent_units": false,
+
+            "redraw": false
+        }
+    },
+    created: function() {
+        // Javascript has the best indirection...
+        var rule = this;
+
+        var request = new XMLHttpRequest();
+
+        request.addEventListener("load", function() {
+            rule.subplans = JSON.parse(request.response);
+
+            rule.check_options();
+        });
+        request.open("GET", "/api/model/subplan/?format=json");
```

Comment:
Might be a good idea to request from my API instead:
`/api/search/?select=id,code,name,year,units&from=subplan&year=subplan_year`

This would mean you could restrict by the year of the Subplan since staff won't be creating degrees for 2019 containing Subplans from other years (You'll obviously have to link `subplan_year` to the actual value). 

It might also be a good idea to add `&code=-MAJ`, `&code=-MIN`, or `&code=-SPEC` to search for only majors, minors, or specialisations, since they aren't supposed to be interchangeable. This would also require some changes to the UI to allow the user to select the plan type

Reply:
```
Might be a good idea to request from my API instead:
/api/search/?select=id,code,name,year,units&from=subplan&year=subplan_year
```
Yeah, I'll take a look down the line, though it isn't critical for this PR.

```
It might also be a good idea to add &code=-MAJ, &code=-MIN, or &code=-SPEC to search for only majors, minors, or specialisations, since they aren't supposed to be interchangeable. This would also require some changes to the UI to allow the user to select the plan type
```
We already discussed this one. A warning does appear, and feel like this isn't all that required IMO. Proper search might be a better alternative.

Reply:
My issue wasn't that there is no warning, it's that we are letting them select mutliple mutually exclusive courses only to tell them they aren't allowed. So why let them select them in the first place? As you've said, the PR still works and passes all criteria, so I'm fine for you to push. Just thought it was something we could work on later